GH-86275: Implementation of hypothesis stubs for property-based tests, with zoneinfo tests

pganssle opened the pull request:
This PR adds the property tests for `zoneinfo` currently maintained at https://github.com/Zac-HD/stdlib-property-tests, along with a compatibility interface for `hypothesis` that gracefully degrades to simple parameterized tests.

I haven't stubbed out the entire interface, but I have stubbed out a bit more than I am using as part of the `zoneinfo` tests right now (mostly the subset that I saw in use in the other `stdlib-property-tests`). We can scale down the stubbed interface if we want, and then have people add in the parts they need as it comes up. I don't think it's a good idea to proactively stub out the entire `hypothesis` interface, since those things would need maintenance over time.

If we like this approach, I will also add in some tooling to run the tests _with_ hypothesis, for use in a buildbot and/or optional PR job.

CC: @Zac-HD, @cfbolz 

<!-- issue-number: [bpo-42109](https://bugs.python.org/issue42109) -->
https://bugs.python.org/issue42109
<!-- /issue-number -->


<!-- gh-issue-number: gh-86275 -->
* Issue: gh-86275
<!-- /gh-issue-number -->

Zac-HD commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,71 @@
+
+from . import strategies
+from ._settings import HealthCheck, Verbosity, settings
+from .control import assume, reject
```
For stubs, I'd just define all of these inline in the `__init__` file - they should only ever be accessed via `hypothesis.foo` (and we should get on with https://github.com/HypothesisWorks/hypothesis/issues/2175) - but this certainly works.

pganssle replied:
Sounds good.

Zac-HD commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,71 @@
+                        f(self, *example_args, **example_kwargs)
+
+        else:
+            test_function = unittest.skip(
```
It would be nice to add a comment here too that `@example` will re-wrap the original function if applied after this `@given` decorator.

pganssle replied:
Done.

Zac-HD reviewed (approved):
Looks great - I'm *very* excited to get this running :grin:

Zac-HD commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,71 @@
+        if getattr(f, "_given", False):
+            # given can be used more than once, but for our purposes there's
+            # no need to wrap the function multiple times.
+            return f
```
Actually, applying `@given()` multiple times is an error, so this can be taken out too.

pganssle replied:
Ah, I didn't realize that had changed!

pganssle commented on Lib/test/test_zoneinfo/test_zoneinfo_property.py:
```diff
@@ -0,0 +1,365 @@
+    raise unittest.SkipTest("No time zone data available")
+
+
+def valid_keys():
```
These tests were implemented before [this](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.timezone_keys), and the latest version of `hypothesis` has a `.timezone_keys` strategy as well as a `.timezones` strategy, so we should probably refactor this, but there's no rush.

pganssle replied:
Upon further review, we can maybe do this in the future. I'm not sure that `timezone_keys` does exactly the same thing as `valid_keys`.

terryjreedy commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+from enum import Enum
```
I think this or hypothesis_helper needs a docstring minimally explaining use and referring back to hypothesis package and doc.

pganssle replied:
Agreed.

terryjreedy replied:
As I later asked, this is still a draft and not claimed to be merge-ready.

terryjreedy commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+from . import strategies
+
+
+def given(*_args, **_kwargs):
```
Why the underscores?

Functions need docstrings.  This one should mentions that arguments are not used as randomized strategies are not run.

pganssle replied:
The underscores are because it's unused. Might be Google-isms or pylint-isms getting to me, because pylint won't complain about unused arguments if you start them with _.

terryjreedy replied:
OK, never seen this.  It might or might not be an issue for others.

terryjreedy commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+                        f(self, *example_args, **example_kwargs)
+
+        else:
+            # If we have found no examples, we must skip the test. If @example
+            # is applied after @given, it will re-wrap the test to remove the
+            # skip decorator.
```
Does hypothesis allow intermixing @given and @examples?  Should suggestion for stdlib use be that @given is outermost (topmost) decorator?

pganssle replied:
Hypothesis allows you to do stuff like:

```python
@example(3)
@given(i=strategies.integers())
@example(4)
def test_i(i):
    ...
```

It's easy enough to support that here, so I see no reason to force people to change the order they want to apply their decorators in.

terryjreedy replied:
OK

terryjreedy commented on Lib/test/support/_hypothesis_stubs/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+# Stub out only the subset of the interface that we actually use in our tests.
+class StubClass:
+    def __init__(self, *args, **kwargs):
+        super().__init__()
```
I don't believe I have ever seen this in direct object subclass.  What does `object.__init__()` do?

pganssle replied:
Presumably nothing, but I believe this is relevant when there is some more complicated MRO.

TBH I can never keep the details straight and I can't remember if I did this because it was necessary or out of habit. Looking at it now, it seems like this is the base of a linear hierarchy, so right now `super().__init__()` will always call `object.__init__()`, but I'm not sure it's *hurting* anything...

terryjreedy replied:
It raises questions in a reviewers/readers mind, such as "Did you forget to subclass from somthing else?"

pganssle replied:
Done.

terryjreedy commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+    "Verbosity",
+]
+
+from . import strategies
```
A linter might recommend removing import as unused.   Add comment such `# Added to __all__ for users of this module.`?

pganssle replied:
Shouldn't the linter notice that it's part of `__all__`, or that we're in `__init__.py`? This is done to expose the `strategies` name eagerly, so that `import hypothesis; hypothesis.strategies.integers` will work. it's fairly idiomatic, so I don't expect anyone to be surprised by it.

terryjreedy replied:
I was, until I looked up instead of down for its use, and saw usage examples in the tests.

terryjreedy commented on Lib/test/test_zoneinfo/test_zoneinfo_property.py:
```diff
@@ -0,0 +1,365 @@
+        # datetimes is always False.
+        if py_dt != c_dt:
+            self.assertEqual(
+                self._is_ambiguous(py_dt), self._is_ambiguous(c_dt), (py_dt, c_dt),
```
Trailing comma is legal but misleading as these are not a series of 'equivalent' args and nothing more can be added.
```suggestion
                self._is_ambiguous(py_dt), self._is_ambiguous(c_dt), (py_dt, c_dt)
```

pganssle replied:
Hm.. I've been maintaining the style with `black` upstream so I think I'd prefer not to change it for trivial reasons, to make merges easier (though this is broken up across multiple lines in the [upstream repo](https://github.com/pganssle/zoneinfo/blob/c54d700fcba78b20733390ee3014809ad69858f6/tests/test_zoneinfo_property.py#L265-L270), so I dunno what's going on there.

terryjreedy replied:
This is not idiomatic in the stdlib I have seen.  I understand your concern with your upstream (as in, I won't object further), but this makes me like black even less ;-).

terryjreedy reviewed (changes requested):
2 PRs in 1.  Do you consider the hypothesis_helper part to be finished and ready to merge?

I intend to add more on the issue later.

pganssle reviewed (commented):
> 2 PRs in 1. Do you consider the hypothesis_helper part to be finished and ready to merge?

The content is probably ready to go, but it's still mostly a draft, so it doesn't have niceties like documentation and NEWS and whatnot. I figured I'd get something that works and allow that to be used as a base for implementing something that works.

I agree this is_sort of_ 2 PRs in one, but we don't have anything that uses hypothesis in the standard library, so I think we need something here in order to exercise the code paths. We could switch it out for a simpler set of hypothesis tests (though presumably a simpler set would exercise a smaller subset of the stubs).

If we break it into multiple PRs, we should do it at the end, immediately before merge.

Zac-HD commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def settings(*args, **kwargs):
+    pass  # pragma: nocover
```
`@hypothesis.settings(...)` can be used as a decorator, so perhaps 

```suggestion
def settings(*args, **kwargs):
    return lambda f: f  # pragma: nocover
```

Zac-HD commented on Lib/test/support/_hypothesis_stubs/__init__.py:
```diff
@@ -0,0 +1,102 @@
+        return list(cls)
+
+
+class Verbosity(Enum):
```
We might also want [the `Phase` enum](https://hypothesis.readthedocs.io/en/latest/settings.html#phases)?

Zac-HD commented on Lib/test/support/_hypothesis_stubs/_helpers.py:
```diff
@@ -0,0 +1,34 @@
+        if kwargstr:
+            in_parens += ", " + kwargstr
+
+        return f"{self.__qualname__}({in_parens})"
```
Since the various `.map()`, `.filter()`, `.__or__()` methods don't alter the repr at all, I'd probably prefer that this be less specific - better to represent `integers().map(str)` as `<stub object at 0x...>` than `integers()`!

I also don't think it's worth the code to make the repr track all the methods, unless adding a `Stub.__trailing_repr` argument would work?

pganssle replied:
OK, this should be working now. Here's a little test file:

```python
from test.support.hypothesis_helper import hypothesis

def complement(x):
    return not x

def identity(x):
    return x

st = hypothesis.strategies

test_strategies = [
    st.integers(),
    st.floats().filter(complement),
    st.lists(st.booleans).flatmap(identity),
    st.characters() | st.dates().map(lambda x: x),
    st.booleans() | st.uuids(),
]

for test_strategy in test_strategies:
    print(repr(test_strategy))
```

Results:

```
hypothesis.strategies.integers()
hypothesis.strategies.floats().filter(complement)
hypothesis.strategies.lists().flatmap(identity)
one_of(hypothesis.strategies.characters(), hypothesis.strategies.dates().map(<lambda>))
one_of(hypothesis.strategies.booleans(), hypothesis.strategies.uuids())
```

pganssle committed:
Add stubs for hypothesis tests

These are stubs to be used for adding hypothesis
(https://hypothesis.readthedocs.io/en/latest/) tests to the standard
library.

When the tests are run in an environment where `hypothesis` and its
various dependencies are not installed, the stubs will turn any tests
with examples into simple parameterized tests and any tests without
examples are skipped.

pganssle committed:
Add examples to zoneinfo hypothesis tests

pganssle committed:
Add property tests for the zoneinfo module

This migrates the tests from
https://github.com/Zac-HD/stdlib-property-tests into the standard
library, using the hypothesis stubs.

pganssle committed:
Enable settings to operate as a decorator

Co-authored-by: Zac Hatfield-Dodds <zac.hatfield.dodds@gmail.com>

pganssle committed:
Add Phase enum

pganssle committed:
Make reprs more accurate

pganssle committed:
Add news entry

Zac-HD commented on Lib/test/test_zoneinfo/test_zoneinfo_property.py:
```diff
@@ -0,0 +1,365 @@
+        resource_name = components[-1]
+
+        try:
+            return resources.is_resource(package_name, resource_name)
```
Tests are failing with a `DeprecationWarning`, [and the docs recommend the new `files()` interface](https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy) e.g. by copying [this backport function](https://github.com/python/importlib_resources/blob/66ea2dc7eb12b1be2322b7ad002cefb12d364dff/importlib_resources/_legacy.py#L60-L69).

pganssle replied:
Done.

pganssle committed:
Add Azure Pipelines CI and PR jobs for hypothesis

pganssle committed:
Add GHA job to run Hypothesis tests

This version of the GHA tries to build CPython once, package it in a
tarball, then re-use it for the normal tests and the hypothesis tests.
It doesn't seem to be working.

It excludes many of the slowest tests manually, plus one test that is
broken for unrelated reasons.

pganssle committed:
Hard-code ignoring hypothesis files in libregrtest

pganssle committed:
Use independent build stages for hypothesis GHA job

hugovk commented on .github/workflows/build.yml:
```diff
@@ -185,7 +197,7 @@ jobs:
 
   build_ubuntu:
-    name: 'Ubuntu'
+    name: 'Build on Ubuntu'
```
This change causes this required check to be never met:

<img width="843" alt="image" src="https://user-images.githubusercontent.com/1324225/235838297-7f42355f-9e14-4fe6-8927-d91ad47bd1eb.png">


Let's keep the original name, otherwise we'll need to sync all the other open PRs:

```suggestion
    name: 'Ubuntu'
```

pganssle replied:
Ah, this was actually an accidental artifact from my original approach, which was to do the build on Ubuntu once, upload it as an artifact, then download it in two separate jobs, one that ran the test suite with hypothesis and one without.

I didn't immediately get it to work and it didn't seem on the critical path so I reverted it, but I guess the name got changed in the shuffle, I've changed it back now.

hugovk commented on .github/workflows/build.yml:
```diff
@@ -291,6 +303,89 @@ jobs:
 
+  test_hypothesis:
+    name: "Hypothesis Tests on Ubuntu"
+    runs-on: ubuntu-20.04
```
Recently added to the other jobs, let's add here too:
```suggestion
    runs-on: ubuntu-20.04
    timeout-minutes: 60
```

pganssle replied:
Good call.

pganssle committed:
Revert "Add Azure Pipelines CI and PR jobs for hypothesis"

This reverts commit 5ba1a6a5a701c333278bce5554d93529177b9984.

hugovk reviewed (approved):
CI changes look good. I agree it's better to use GitHub Actions than Azure Pipelines, it's more familiar and easier to restart failures.

The new `test_hypothesis` job has a lot of duplication with `build_ubuntu`. It would be nice to deduplicate this in the future, possibly via a re-usable workflow, but this PR is old enough that it would be good to get it merged.

Thanks!
